presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -132,14 +138,30 @@
     private final ExecutionPolicy executionPolicy;
     private final SubPlan plan;
     private final StreamingPlanSection sectionedPlan;
+
+    // this map is synchronized and will get updated when stages retry
     private final Map<StageId, StageExecutionAndScheduler> stageExecutions;
+
     private final ExecutorService executor;
     private final StageId rootStageId;
     private final SplitSchedulerStats schedulerStats;
     private final boolean summarizeTaskInfo;
     private final AtomicBoolean started = new AtomicBoolean();
     private final AtomicBoolean scheduling = new AtomicBoolean();
     private final int maxConcurrentMaterializations;
+    private final int maxStageRetries;
+    private final AtomicInteger retriedSections = new AtomicInteger();
+    private final Metadata metadata;
```

Comment:
These fields (from `metadata` to `nodeTaskMap`)  are added to support retry.  Maybe add one empt line before this and a comment ? :) 

I am also wondering whether we should encapsulate them into something like `StageExecutionFactory`.. ?